File: typescript/src/endpoints/filings.ts
/**
 * Filing endpoints for SEC EDGAR API
 */

import { HttpClient } from "../utils";
import { CompanySubmissions, RequestOptions } from "../types";

const SEC_BASE_URL = "https://data.sec.gov";
const SEC_ARCHIVES_URL = "https://www.sec.gov";
const SEC_CURRENT_FILINGS_URL = "https://www.sec.gov/cgi-bin/browse-edgar";

export interface RecentFilingEntry {
  cik: string;
  accessionNumber: string;
  formType: string;
  filingDate: string;
  companyName: string;
  url: string;
}

export class FilingsEndpoints {
  private httpClient: HttpClient;

  constructor(httpClient: HttpClient) {
    this.httpClient = httpClient;
  }

  async getCompanySubmissions(
    cik: string | number,
    options: RequestOptions = {},
  ): Promise<CompanySubmissions> {
    const cikStr =
      typeof cik === "number"
        ? cik.toString().padStart(10, "0")
        : cik.padStart(10, "0");
    const url = `${SEC_BASE_URL}/submissions/CIK${cikStr}.json`;

    let submissions = await this.httpClient.get(url);

    // Apply filters if provided
    if (options.submissionType || options.fromDate || options.toDate) {
      submissions = this.filterSubmissions(submissions, options);
    }

    return submissions;
  }

  async getFiling(
    cik: string | number,
    accessionNumber: string,
  ): Promise<Record<string, any>> {
    // Archive folders live on www.sec.gov, use the unpadded CIK, and the
    // folder listing is served as index.json
    const cikStr = typeof cik === "number" ? cik.toString() : cik;
    const accessionFormatted = accessionNumber.replace(/-/g, "");
    const url = `${SEC_ARCHIVES_URL}/Archives/edgar/data/${parseInt(cikStr, 10)}/${accessionFormatted}/index.json`;

    return await this.httpClient.get(url);
  }

  /**
   * Fetch one older-history submissions page (from `filings.files`).
   *
   * The main submissions payload lists additional filing history in
   * `filings.files` (e.g. `CIK0000320193-submissions-001.json`); this
   * fetches one of those pages. The returned payload has the same
   * columnar shape as `filings.recent`.
   */
  async getCompanySubmissionsPage(
    pageName: string,
  ): Promise<Record<string, any>> {
    const url = `${SEC_BASE_URL}/submissions/${pageName}`;
    return await this.httpClient.get(url);
  }

  /**
   * Get recent filings across all companies via the SEC EDGAR Atom feed.
   *
   * @param formType Form type(s) to filter by (e.g., "8-K", ["3", "4", "5"])
   * @param limit Maximum number of filings to return (feed max 100/page)
   * @param owner Ownership-filing filter: "include", "exclude", or "only"
   * @param start Pagination offset
   */
  async getRecentFilings(
    formType?: string | string[],
    limit: number = 40,
    owner: "include" | "exclude" | "only" = "include",
    start: number = 0,
  ): Promise<RecentFilingEntry[]> {
    const formTypes =
      formType == null
        ? [undefined]
        : Array.isArray(formType)
          ? formType
          : [formType];

    const entries: RecentFilingEntry[] = [];
    for (const type of formTypes) {
      const params = new URLSearchParams({
        action: "getcurrent",
        output: "atom",
        owner,
        count: String(Math.min(limit, 100)),
        start: String(start),
      });
      if (type) {
        params.set("type", type);
      }
      try {
        const feed = await this.httpClient.getRaw(
          `${SEC_CURRENT_FILINGS_URL}?${params.toString()}`,
        );
        entries.push(...FilingsEndpoints.parseAtomFeed(feed));
      } catch {
        // A failed feed request degrades to an empty page for that type
      }
    }

    entries.sort((a, b) =>
      a.filingDate < b.filingDate ? 1 : a.filingDate > b.filingDate ? -1 : 0,
    );
    return entries.slice(0, limit);
  }

  /** Parse the browse-edgar Atom feed into filing entries. */
  static parseAtomFeed(feed: string): RecentFilingEntry[] {
    const entries: RecentFilingEntry[] = [];
    const entryPattern = /<entry>([\s\S]*?)<\/entry>/g;
    let entryMatch;
    while ((entryMatch = entryPattern.exec(feed)) !== null) {
      const block = entryMatch[1];
      const title = (block.match(/<title>([\s\S]*?)<\/title>/) || [])[1] || "";
      // Title format: "10-K - Company Name (0001234567) (Filer)"
      const titleMatch = title.match(/^\s*(.+?)\s+-\s+(.*?)\s+\((\d{10})\)/);
      const link = (block.match(/<link[^>]*href="([^"]+)"/) || [])[1] || "";
      const accession =
        (block.match(/accession[-_]?number[^>]*>([\d-]+)</i) || [])[1] ||
        (link.match(/(\d{10}-\d{2}-\d{6})/) || [])[1] ||
        "";
      const updated = (block.match(/<updated>([\d-]+)/) || [])[1] || "";

      if (titleMatch) {
        entries.push({
          formType: titleMatch[1].trim(),
          companyName: titleMatch[2].trim(),
          cik: titleMatch[3],
          accessionNumber: accession,
          filingDate: updated,
          url: link,
        });
      }
    }
    return entries;
  }

  private filterSubmissions(
    submissions: CompanySubmissions,
    options: RequestOptions,
  ): CompanySubmissions {
    if (!submissions.filings || !submissions.filings.recent) {
      return submissions;
    }

    const { recent } = submissions.filings;
    let indices = recent.accessionNumber.map((_: any, index: number) => index);

    // Filter by submission type
    if (options.submissionType) {
      indices = indices.filter(
        (i: number) => recent.form[i] === options.submissionType,
      );
    }

    // Filter by date range
    if (options.fromDate || options.toDate) {
      indices = indices.filter((i: number) => {
        const filingDate = recent.filingDate[i];
        if (options.fromDate && filingDate < options.fromDate) return false;
        if (options.toDate && filingDate > options.toDate) return false;
        return true;
      });
    }

    // Create filtered results
    const filteredRecent: Record<string, any> = {};
    Object.keys(recent).forEach((key) => {
      filteredRecent[key] = indices.map((i: number) => recent[key][i]);
    });

    return {
      ...submissions,
      filings: {
        ...submissions.filings,
        recent: filteredRecent,
      },
    };
  }
}

File: python/src/sec_edgar_toolkit/endpoints/filings.py
"""SEC filing endpoints."""

from __future__ import annotations

import logging
import re
import xml.etree.ElementTree as ET
from typing import Any, Dict, List, Optional, Union
from urllib.parse import urljoin

from ..utils import FilingFilter, HttpClient

logger = logging.getLogger(__name__)


class FilingsEndpoints:
    """
    Endpoints for SEC filing data and submissions.

    This class handles all filing-related API endpoints including:
    - Company submissions (all filings for a company)
    - Individual filing details
    - Filing document retrieval
    - Recent filings across all companies via EDGAR RSS
    - Filtering and search capabilities

    Args:
        http_client: HTTP client instance for making requests

    Example:
        >>> client = HttpClient("MyApp/1.0 (contact@example.com)")
        >>> endpoints = FilingsEndpoints(client)
        >>> submissions = endpoints.get_company_submissions("0000320193")
    """

    DATA_URL = "https://data.sec.gov/"
    ARCHIVES_URL = "https://www.sec.gov/"
    CURRENT_FILINGS_URL = "https://www.sec.gov/cgi-bin/browse-edgar"

    def __init__(self, http_client: HttpClient) -> None:
        """Initialize filings endpoints."""
        self.http_client = http_client

    def get_company_submissions(
        self,
        cik: Union[str, int],
        submission_type: Optional[str] = None,
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Get all submissions for a specific company.

        This endpoint returns detailed company information along with
        all their filing history.

        Args:
            cik: Company CIK number
            submission_type: Filter by form type (e.g., '10-K', '10-Q')
            from_date: Start date (YYYY-MM-DD format)
            to_date: End date (YYYY-MM-DD format)

        Returns:
            Company submissions data including all filings

        Raises:
            NotFoundError: If company CIK is not found

        Example:
            >>> subs = endpoints.get_company_submissions("0000320193")
            >>> print(f"Company: {subs['name']}")
            >>> print(f"Total filings: {len(subs['filings']['recent'])}")
        """
        # Normalize CIK to 10-digit string
        cik_str = str(cik).zfill(10)

        url = urljoin(self.DATA_URL, f"submissions/CIK{cik_str}.json")
        data = self.http_client.get(url)

        return self._filter_submissions(data, submission_type, from_date, to_date)

    def get_company_submissions_page(self, page_name: str) -> Dict[str, Any]:
        """
        Fetch one older-history submissions page.

        The main submissions payload lists additional filing history in
        ``filings.files`` (e.g. ``CIK0000320193-submissions-001.json``);
        this fetches one of those pages. The returned payload has the
        same columnar shape as ``filings.recent``.

        Args:
            page_name: The ``name`` entry from ``filings.files``

        Returns:
            Columnar filings data (``accessionNumber``, ``form``, ...)
        """
        url = urljoin(self.DATA_URL, f"submissions/{page_name}")
        return self.http_client.get(url)

    def _filter_submissions(
        self,
        data: Dict[str, Any],
        submission_type: Optional[str],
        from_date: Optional[str],
        to_date: Optional[str],
    ) -> Dict[str, Any]:

        # Apply filters if provided
        if submission_type or from_date or to_date:
            filings_data = data.get("filings", {})
            if isinstance(filings_data, dict) and "recent" in filings_data:
                filtered_filings = FilingFilter.filter_filings(
                    filings_data["recent"],
                    submission_type,
                    from_date,
                    to_date,
                )
                filings_data["recent"] = filtered_filings

        return data

    def get_filing(
        self,
        cik: Union[str, int],
        accession_number: str,
    ) -> Dict[str, Any]:
        """
        Get specific filing details and documents.

        Args:
            cik: Company CIK number
            accession_number: Accession number of the filing

        Returns:
            Filing details including document list

        Example:
            >>> filing = endpoints.get_filing("0000320193", "0000320193-23-000077")
            >>> print(f"Form type: {filing['form']}")
            >>> print(f"Filed on: {filing['filingDate']}")
        """
        # Archive folders live on www.sec.gov and use the unpadded CIK;
        # the folder listing is served as index.json
        cik_str = str(int(str(cik)))
        accession = accession_number.replace("-", "")

        url = urljoin(
            self.ARCHIVES_URL,
            f"Archives/edgar/data/{cik_str}/{accession}/index.json",
        )

        return self.http_client.get(url)

    def get_recent_filings(
        self,
        form_type: Optional[Union[str, List[str]]] = None,
        limit: int = 40,
        owner: str = "include",
        start: int = 0,
    ) -> List[Dict[str, Any]]:
        """
        Get recent filings across all companies using the SEC EDGAR Atom feed.

        This endpoint queries the SEC EDGAR current filings feed to retrieve
        recently filed documents without requiring a specific company identifier.

        Args:
            form_type: Filter by form type(s) (e.g., "10-K", ["10-K", "10-Q"]).
                       When a list is provided, multiple requests are made and
                       results are merged.
            limit: Maximum number of filings to return (default: 40, max: 100)
            owner: Ownership filter - "include", "exclude", or "only" (default: "include")
            start: Starting offset for pagination (default: 0)

        Returns:
            List of filing dicts with keys: cik, accession_number, form_type,
            filing_date, company_name, url

        Example:
            >>> filings = endpoints.get_recent_filings(form_type="10-K", limit=10)
            >>> for f in filings:
            ...     print(f"{f['company_name']}: {f['form_type']} ({f['filing_date']})")
        """
        if isinstance(form_type, list):
            # Merge results from multiple form types
            all_filings: List[Dict[str, Any]] = []
            per_type_limit = max(limit // len(form_type), 10)
            for ft in form_type:
                type_filings = self._fetch_current_filings_feed(
                    form_type=ft, count=per_type_limit, owner=owner, start=start
                )
                all_filings.extend(type_filings)

            # Sort by filing date descending and trim to limit
            all_filings.sort(key=lambda x: x.get("filing_date", ""), reverse=True)
            return all_filings[:limit]

        return self._fetch_current_filings_feed(
            form_type=form_type, count=min(limit, 100), owner=owner, start=start
        )

    def _fetch_current_filings_feed(
        self,
        form_type: Optional[str] = None,
        count: int = 40,
        owner: str = "include",
        start: int = 0,
    ) -> List[Dict[str, Any]]:
        """
        Fetch recent filings from the SEC EDGAR Atom feed.

        Args:
            form_type: Form type filter (e.g., "10-K")
            count: Number of results to fetch
            owner: Ownership filter
            start: Pagination offset

        Returns:
            List of filing dicts parsed from the Atom feed
        """
        params = {
            "action": "getcurrent",
            "type": form_type or "",
            "dateb": "",
            "owner": owner,
            "count": str(count),
            "search_text": "",
            "start": str(start),
            "output": "atom",
        }

        try:
            content = self.http_client.get_raw(self.CURRENT_FILINGS_URL, params=params)
            return self._parse_atom_feed(content)
        except Exception as e:
            logger.error(f"Failed to fetch current filings feed: {e}")
            return []

    @staticmethod
    def _parse_atom_feed(content: bytes) -> List[Dict[str, Any]]:
        """
        Parse an Atom feed of SEC filings into structured dicts.

        Args:
            content: Raw XML bytes of the Atom feed

        Returns:
            List of filing dicts with keys: cik, accession_number, form_type,
            filing_date, company_name, url
        """
        ns = {"atom": "http://www.w3.org/2005/Atom"}
        filings: List[Dict[str, Any]] = []

        try:
            root = ET.fromstring(content)
        except ET.ParseError as e:
            logger.error(f"Failed to parse Atom feed: {e}")
            return filings

        for entry in root.findall("atom:entry", ns):
            try:
                title_el = entry.find("atom:title", ns)
                title = (title_el.text or "") if title_el is not None else ""

                link_el = entry.find("atom:link", ns)
                url = link_el.get("href", "") if link_el is not None else ""

                summary_el = entry.find("atom:summary", ns)
                summary = (
                    summary_el.text.strip()
                    if summary_el is not None and summary_el.text
                    else ""
                )

                category_el = entry.find("atom:category", ns)
                form = category_el.get("term", "") if category_el is not None else ""

                # Parse CIK from title: "10-K - Company Name (0000946644) (Filer)"
                cik = ""
                cik_match = re.search(r"\((\d{7,10})\)", title)
                if cik_match:
                    cik = cik_match.group(1).zfill(10)

                # Parse accession number from summary: "AccNo: 0001493152-26-013301"
                accession = ""
                acc_match = re.search(r"AccNo.*?(\d{10}-\d{2}-\d{6})", summary)
                if acc_match:
                    accession = acc_match.group(1)

                # Parse filing date from summary: "Filed: 2026-03-27"
                filing_date = ""
                date_match = re.search(r"Filed.*?(\d{4}-\d{2}-\d{2})", summary)
                if date_match:
                    filing_date = date_match.group(1)

                # Parse company name from title: "10-K - Company Name (CIK) (Filer)"
                company_name = ""
                name_match = re.match(r"[\w\-/]+ - (.+?)\s*\(\d+\)", title)
                if name_match:
                    company_name = name_match.group(1).strip()

                filings.append(
                    {
                        "cik": cik,
                        "accession_number": accession,
                        "form_type": form,
                        "filing_date": filing_date,
                        "company_name": company_name,
                        "url": url,
                    }
                )
            except Exception as e:
                logger.debug(f"Failed to parse feed entry: {e}")
                continue

        return filings
